mesonbuild/build.py

```diff
@@ -362,6 +363,26 @@ def get_project_link_args(self, compiler: 'Compiler', project: str, for_machine:
 
         return link_args.get(compiler.get_language(), [])
 
+@unique
+class IncludeDirsUse(Enum):
+    # 'incdirs', if not absolute paths, are to be used as both src-relative and
+    # build-relative.  Absolute paths are allowed and will just result in one
+    # include dir.
+    src_and_build_relative = 0
+
+    # 'incdirs' will only be used as src-relative (or absolute) include directories.
+    src_relative = 1
+
+    # 'incdirs' are expected to be build-relative include dirs and will error if
+    # absolute paths are used.
+    build_relative = 2
```

Comment:
In general we use `ALL_CAPS` names for enum fields, and I'd prefer to keep it that way.

These are also really verbose names, could we just have `SRC`, `BUILD`, `BOTH`?

Reply:
I think that an enum is very wrong here, because it fundamentally ties to the meson.build API contract being offered, which is string-based and should not be.

Reply:
I'd expect an API at the meson.build level to look like this:

```meson
incdirs = include_directories('headers', build: false, source: true)
```

This is much better than:
```meson
incdirs: include_directories('headers', use_type: 'src-and-build-relative')
```

It's easier for users to understand, and it's also easier for the interpreter to validate because we don't need an enum of magic strings, we just need two boolean properties...

Reply:
Personally, I prefer the two bools option. Having both values to `false` is not a problem, as you can always add a validator an issue and error if this is the case.

Reply:
Since it's not too a big deal either way and you say you prefer the two bools style in response to my points above, I've added a [separate commit](https://github.com/mesonbuild/meson/pull/11984/commits/b162eb46c0e5c83085da29f225e93fdbb971c4ac) for now (just in case anyone wants to compare the two alternatives) that uses that approach.  This approach keeps 2 bools throughout the internal structures and so, as mentioned above, needs to double up the validation for the invalid combination that this style now invites.